timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/ExclusiveLock.java

```diff
@@ -108,4 +95,35 @@ private void checkIsCurrentHolder(UUID requestId) {
         }
     }
 
+    @NotThreadSafe
+    private static class LockRequestQueue {
+
+        @SuppressWarnings("checkstyle:illegaltype")
+        private final LinkedHashMap<UUID, LockRequest> queue = Maps.newLinkedHashMap();
+
+        public void enqueue(LockRequest request) {
+            LockRequest existingRequest = queue.put(request.requestId, request);
+            if (existingRequest != null) {
+                queue.put(request.requestId, existingRequest);
+                throw new LoggableIllegalStateException(
+                        "Cannot enqueue the same request id twice.",
+                        SafeArg.of("requestId", request.requestId));
+            }
```

Comment:
Hmm, does this send the existingRequest to the back of the iteration order? I'm not sure if we have any guarantees on this or how the LinkedHashMap impl works.

That said even if it does this is probably a super rare issue (I'm guessing we catch retried requests somewhere above)